Framework/Algorithms/src/NormaliseToMonitor.cpp

```diff
@@ -115,7 +115,14 @@ bool MonIDPropChanger::monitorIdReader(
   }
   // are these monitors really there?
   // got the index of correspondent spectra.
-  std::vector<size_t> indexList = inputWS->getIndicesFromDetectorIDs(mon);
+  // Only check if number of histograms is small, else this takes too long!
+  std::vector<size_t> indexList;
+  if (inputWS->getNumberHistograms() < 100000) {
+    indexList = inputWS->getIndicesFromDetectorIDs(mon);
+  } else {
+    indexList = std::vector<size_t>(mon.begin(), mon.end());
```

Comment:
Not sure I understand this. This seems to be setting `workspace index = detector ID`?

Reply:
Hmm, looking into this more closely it gets confusing. This is meant to be a list of monitor IDs, so I think the else part is fine. In a previous update I made some changes to this, but it looks like if monitors are missing from the workspace then the vector is reduced in size, but does not correspond to what is actually present?

https://github.com/mantidproject/mantid/pull/20804/files#diff-f54358cb406d12adce22fb09ef086959L124

Reply:
I will see if I can rework this to something sensible. I am also tempted to remove the check as it covers a fairly uncommon case.

Reply:
I am not sure why you think the `else` part is fine. `getIndicesFromDetectorIDs` translates from detectors IDs to workspace indices. "monitor IDs" is not a defined concept, so I am not sure what you are referring to?

Reply:
To be clear when I mention monitor IDs I mean detector IDs (for detectors that are monitors). This how it has been defined in the `NormaliseToMonitor` docs - see [`MonitorID`](http://docs.mantidproject.org/nightly/algorithms/NormaliseToMonitor-v1.html). I can make the documentation a little clearer, but I am afraid we are stuck with the name until a v2 of the algorithm comes along.

The idea in this bit of code is to get a list of the monitor (detector) IDs in the workspace. First the monitor IDs are obtained from the instrument, then a check is made to see if corresponding spectra exist in the workspace. I skip this second check as it is slow on large workspaces.

Reply:
What I am saying is that this is not just doing a check, but also a translation to workspace index. You seem to be skipping that translation when skipping the check.

Reply:
I'll also rename this from `indexList`, the name is just wrong.

Reply:
Yes, but the only the size of the translated vector is used, not the contents.

Reply:
Hmm, indeed, how weird. Where does it find the corresponding spectrum then?

Reply:
I don't see how this would ever have worked correctly before. I have pushed some updates that should improve things, and I think the performance is good enough to remove the check on the number of spectra.

Reply:
It can be tested with:
```python
CreateSampleWorkspace(OutputWorkspace='big1', NumBanks=100, NumMonitors=10, BankPixelWidth=100, NumScanPoints=10)
ExtractSpectra(InputWorkspace='big1', OutputWorkspace='extracted_mons', StartWorkspaceIndex=10000090, EndWorkspaceIndex=10000099)
```

Reply:
I will mark this as in progress and do some tidy up tomorrow.

Reply:
@SimonHeybrock I have pushed the changes, this should work as expected now. You can try the script, then open the dialogue from MantidPlot. It is still a little slow on this workspace, but I think the performance is sufficient to remove the checks on histogram numbers. In most real cases it will now be quick to open now, as only the first spectra are checked if all the monitors are found.